### DISCUSSION TITLE
Private CircleCI Orb upgrade fail

### DISCUSSION THREAD

**Discussion by jBouyoud** [renovatebot/renovate#21718] [Request Help]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the bug

Hi,

We used renovate to manage upgrade of our circleci private orb in pipelines. 
Since a few times we got a warning on renovate dashboard.

Looks like the authentication is broken.

> ⚠ Dependency Lookup Warnings ⚠
> Renovate failed to look up the following dependencies: private-orb-name.
> Files affected: .circleci/config.yml

While this is still working for public Orb.

The actual datasource is based on a `https://circleci.com/graphql-unstable`
I'm not able to find any documentation on a Graphql API at Circleci.
Ref. : https://github.com/renovatebot/renovate/blob/143c9a67516180e364e7a0881b1d9ac4d3fb1b17/lib/modules/datasource/orb/index.ts#L43

Our configs looks like the guide : 
```
  "hostRules": [
    {
      "matchHost": "circleci.com",
      "authType": "Token-Only",
      "encrypted": {
        "token": "***"
      }
    }
  ]
```

I'll be glad to help fixing this issue, please let me known how to proceed.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Failed to look up orb dependency platform (platform)(packageFile=".circleci/config.yml", dependency="platform")
...
    "circleci": [
      {
        "packageFile": ".circleci/config.yml",
        "deps": [
          {
            "depType": "orb",
            "depName": "platform",
            "currentValue": "2",
            "datasource": "orb",
            "packageName": "swile/platform",
            "commitMessageTopic": "{{{depName}}} orb",
            "versioning": "npm",
            "rangeStrategy": "pin",
            "depIndex": 0,
            "updates": [],
            "warnings": [
              {
                "topic": "platform",
                "message": "Failed to look up orb dependency platform"
              }
            ]
          },
...
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by rarkins** [comment]
You will need to find a curl command which works to retrieve your private package from circleci. Maybe the URL changed for example

**Comment by jBouyoud** [comment]
Based on actual circleci CLI code, something like : 

```
>> variables: map[orbVersionRef:my-private-namespace/my-orb@volatile]
>> query: query($orbVersionRef: String!) {
			    orbVersion(orbVersionRef: $orbVersionRef) {
			        id
                                version
                                orb {
                                    id
                                    createdAt
									name
									namespace {
									  name
									}
                                    categories {
                                      id
                                      name
                                    }
	                            statistics {
		                        last30DaysBuildCount,
		                        last30DaysProjectCount,
		                        last30DaysOrganizationCount
	                            }
                                    versions(count: 200) {
                                        createdAt
                                        version
                                    }
                                }
                                source
                                createdAt
			    }
		      }
<< request id: f895d7d7-7acc-4c5b-b45e-85976be91080
<< result status: 200 OK
```

Ref. : https://github.com/CircleCI-Public/circleci-cli/blob/main/api/api.go#L1462

**Comment by rarkins** [comment]
Does Renovate still work for public orbs? We should make the issue title more descriptive once more information is known.

Next step is:
- is the URL (e.g. unstable) still correct?
- is our existing query needing updating?

**Comment by jBouyoud** [comment]
> Does Renovate still work for public orbs?

Yes

> is the URL (e.g. unstable) still correct? 

Yes, `https://circleci.com/graphql-unstable`
And authentication is always handled with `Authentication` header with token as value.

> is our existing query needing updating?

It seems, but unfortunately, I'm not really fluent with GraphQL.

I think you will be more information about the actual error by logging this error when the request fail (https://github.com/renovatebot/renovate/blob/143c9a67516180e364e7a0881b1d9ac4d3fb1b17/lib/modules/datasource/orb/index.ts#LL54C12-L54C12)

**Comment by jBouyoud** [comment]
@rarkins Thanks for that. 

I wait the next deployment to find more details about the error and come you back with that.

**Comment by jBouyoud** [comment]
Unfortunately I'm not able to see you change in `app.renovatebot.com/dashboard`, and still have the same log even the renovate version includes your recent change :

> INFO: Repository started
>   {
>     "renovateVersion": "34.118.3"
>   }
> ...
> DEBUG: Failed to look up orb dependency platform (xyz-orb)(packageFile=".circleci/config.yml", dependency="xyz-orb")

**Comment by rarkins** [comment]
In that case this will require you or someone else to run Renovate in debug mode locally to see what response or error is being returned

**Comment by jBouyoud** [comment]
I'm not able to see any change because the warning log is produce [here](https://github.com/renovatebot/renovate/blob/fabb1bba0828c11c81513958ef3a9226880d6c76/lib/workers/repository/process/lookup/index.ts#L89).

It seems like `Failed to look up orb dependency` is never called and rely on caching mechanism. Since there is no trace of it in `httpstatistics`.

_On a local run :_ 

I build a test repository with the existing config and start renovate : 
`LOG_LEVEL=debug renovate --token=${GITHUB_TOKEN} --dry-run  my-org/debug-renovate`

I reproduce the same behavior. 
Then, I change my config since I original config rely on encrypted token for Renovate Github App. 
Restart and reproduce the same behavior.

By changing the cache directory (`--cache-dir`) works again. 

It is possible that the cache contains some invalid entries?

On my side this is a totally legit scenario since we recently needs to renew all CircleCI tokens (with CircleCI leaks issues). So definitively we have some our repositories not yet up-to-date and rely on an hold token that will respond an unknown response.

I spot multiple levels of caching for this dependency (but I really not relevant to tell where something what's wrong) : 
- [getRawReleases](https://github.com/renovatebot/renovate/blob/fabb1bba0828c11c81513958ef3a9226880d6c76/lib/modules/datasource/index.ts#L320) : L334 ; there is no `if res ..` before `memCache.set(cacheKey, promisedRes);` 
- [getRegistryReleases](https://github.com/renovatebot/renovate/blob/fabb1bba0828c11c81513958ef3a9226880d6c76/lib/modules/datasource/index.ts#LL56C26-L56C26) : L78; seems like Orb datasource never provide `isPrivate``
- [OrbDatasource](https://github.com/renovatebot/renovate/blob/fabb1bba0828c11c81513958ef3a9226880d6c76/lib/modules/datasource/orb/index.ts#L31)

I try to add `isPrivate: !packageName.startsWith('circleci')` in [OrbDatasource](https://github.com/renovatebot/renovate/blob/fabb1bba0828c11c81513958ef3a9226880d6c76/lib/modules/datasource/orb/index.ts#L68) , but this change nothing the same behavior remain.

**Comment by rarkins** [comment]
Once the result is cached, it's cached. The `isPrivate` part you found should help to stop it from being cached _the next time it expires_.

Is there any way I can emulate this myself, or does it require me to have a paid circleci plan? Or for example can you generate a temporary token which is fairly "isolated" and share it with me to debug with? Totally up to you of course. Right now my assumption is that I cannot reproduce this on my side.

**Comment by jBouyoud** [comment]
I just send you an invite on a private repo to be able to test it 🤞

**Comment by rarkins** [comment]
Debugging it, we get an empty result from the GraphQL query.

Here's the query:

```
{
    "query": "\nquery($packageName: String!) {\n  orb(name: $packageName) {\n    name,\n    homeUrl,\n    versions {\n      version,\n      createdAt\n    }\n  }\n}\n",
    "variables": {
        "packageName": "jbouyoud/test-orb"
    }
}
```

Either:
- The query needs to be different for private orbs, or
- There's something peculiar about your orbs

Doesn't sound like the second, so more likely the first

**Comment by jBouyoud** [comment]
With your query there is no difference between a private Orb or a public one.

By running the query with an invalid token, I got the following response : 
```
{
  "data": {
    "orb": null
  }
}
```
I also got the same response for an un-existing Orb.

By running the following query : (add `isPrivate` to Orb struct, found [here](https://github.com/CircleCI-Public/circleci-cli/blob/main/api/api.go#LL765C3-L765C3)) 
```
{
  "query": "\nquery($packageName: String!) {\n  orb(name: $packageName) {\n    name,\n  isPrivate,  homeUrl,\n    versions {\n      version,\n      createdAt\n    }\n  }\n}\n",
  "variables": {
    "packageName": "jbouyoud/test-orb"
  }
}
```

got 

```
{
  "data": {
    "orb": {
      "name": "jbouyoud/test-orb",
      "isPrivate": true,
      "homeUrl": "",
      "versions": [
        {
          "version": "1.0.1",
          "createdAt": "2023-02-02T16:25:57.626Z"
        },
        {
          "version": "1.0.0",
          "createdAt": "2023-02-02T16:25:40.682Z"
        }
      ]
    }
  }
}
```

and for a public one ( `circleci/aws-cli`) got : 
```
{
  "data": {
    "orb": {
      "name": "circleci/aws-cli",
      "isPrivate": false,
      "homeUrl": "",
      "versions": [
        {
          "version": "3.1.4",
          "createdAt": "2022-12-02T01:13:30.190Z"
        },
        {
          "version": "3.1.3",
          "createdAt": "2022-08-25T22:37:42.617Z"
        },
        {
          "version": "3.1.2",
          "createdAt": "2022-08-24T19:41:51.877Z"
        },
        {
          "version": "3.1.1",
          "createdAt": "2022-05-04T07:34:05.336Z"
        },
        {
          "version": "3.1.0",
          "createdAt": "2022-04-21T20:42:06.283Z"
        },
        {
          "version": "3.0.0",
          "createdAt": "2022-04-05T20:27:38.554Z"
        },
        {
          "version": "2.1.0",
          "createdAt": "2022-02-17T23:48:09.401Z"
        },
        {
          "version": "2.0.6",
          "createdAt": "2021-11-19T14:43:21.282Z"
        },
        {
          "version": "2.0.5",
          "createdAt": "2021-11-18T16:49:59.512Z"
        },
        {
          "version": "2.0.4",
          "createdAt": "2021-11-18T15:04:48.995Z"
        }
      ]
    }
  }
}
```

**Comment by rarkins** [comment]
So adding `isPrivate` to the query seems to fix private orbs while not breaking public?

**Comment by jBouyoud** [comment]
It looks like yes 👍

**Comment by rarkins** [comment]
It didn't work for me locally when running against your test repo, but I have pushed the change in a PR anyway

**Comment by jBouyoud** [comment]
Unfortunately, there is no improvements at this point. Seems renovate keep an invalid result in his cache

**Comment by rarkins** [comment]
It never worked for me locally either, without cache. Caches expire too.

**Comment by jBouyoud** [comment]
After renew all config with a new token in all repositories of our organization, I still experience the issue. 
I've tested the token with a circleci cli and I'm able to find my private orb.

And nothing really relevant on logs when request is not cached : 
```
DEBUG: Adding token authentication for circleci.com to hostRules
...
DEBUG: Failed to look up orb organization/orb-name
{
  "res": {
    "data": {
      "orb": null
    }
  }
}
DEBUG: Failed to look up orb dependency orb-name (orb-name)(packageFile=".circleci/config.yml", dependency="orb-name")
...
DEBUG: http statistics
...
 "https://circleci.com/graphql-unstable (POST,200)": 1,
...
 "circleci.com": {
      "requestCount": 1,
      "requestAvgMs": 7629,
      "queueAvgMs": 1
    },
...
```

No more idea on my side 🤷

**Comment by jBouyoud** [comment]
@rarkins any other ideas to fix that ?

**Comment by rarkins** [comment]
No, it needs someone who can reproduce the issue locally to then debug it

**Comment by jBouyoud** [comment]
Hi @rarkins,

I run a lot of tests on a dedicated repo and a dedicated private orb. 
And it's works fine on local renovate run and with OS Renovate Application (with a valid token).
When I put an invalid configuration in hostRules the issue appears and take some time to disappear (or when I clean the cache on local run).

So I still encours the issue with OS Renovate Application and an existing Orb.
I double check all repositories on my organization and all of them are using the same config (which seems valid).
I carefully monitored `Dependency dashboard` during a full day (and sometimes in the day the issue disappear).

My only hint is someone(thing) use renovate with an invalid token for this dependency and block others users (since invalid result/ no results are put in cache).

**Reply by evilensky** [reply]
Hi @jBouyoud . Were you able to get this working? I followed the documentation precisely but can't seem to get successful lookups.  Using hosted mend renovate app.

**Comment by bjaglin-swile** [comment]
Could we be running into per-IP rate limiting, due to a large number of repos using the same token every hour or so? That would explain why you can't reproduce it locally.

[CircleCI is not really clear about their policy](https://circleci.com/docs/api-developers-guide/#rate-limits) (which is understandable), but it follows HTTP standards, so it looks like honoring `Retry-After` as suggested in https://github.com/renovatebot/renovate/issues/20601 could improve this.
